client/state/imports/reducer.js

```diff
@@ -1,13 +1,173 @@
+/**
+ * External dependencies
+ */
+import { get, isEmpty, map, omit, omitBy } from 'lodash';
+
 /**
  * Internal dependencies
  */
-import { combineReducers, withStorageKey } from 'calypso/state/utils';
-import uploadsReducer from 'calypso/state/imports/uploads/reducer';
-import siteImporterReducer from 'calypso/state/imports/site-importer/reducer';
+import {
+	IMPORTS_AUTHORS_SET_MAPPING,
+	IMPORTS_AUTHORS_START_MAPPING,
+	IMPORTS_IMPORT_CANCEL,
+	IMPORTS_IMPORT_LOCK,
+	IMPORTS_IMPORT_RECEIVE,
+	IMPORTS_IMPORT_RESET,
+	IMPORTS_IMPORT_START,
+	IMPORTS_IMPORT_UNLOCK,
+	IMPORTS_START_IMPORTING,
+	IMPORTS_UPLOAD_FAILED,
+	IMPORTS_UPLOAD_COMPLETED,
+	IMPORTS_UPLOAD_SET_PROGRESS,
+	IMPORTS_UPLOAD_START,
+} from 'calypso/state/action-types';
+import { combineReducers, keyedReducer, withStorageKey } from 'calypso/state/utils';
+import uploads from './uploads/reducer';
+import siteImporter from './site-importer/reducer';
+import { appStates } from './constants';
+import { fromApi } from './api';
+
+function isImporterStatusHydrated( state = false, action ) {
+	switch ( action.type ) {
+		case IMPORTS_IMPORT_RECEIVE:
+			return true;
+	}
+
+	return state;
+}
+
+function importerStatus( state = {}, action ) {
+	switch ( action.type ) {
+		case IMPORTS_IMPORT_CANCEL:
+		case IMPORTS_IMPORT_RESET:
+			return omit( state, action.importerId );
+
+		case IMPORTS_UPLOAD_FAILED:
+			return {
+				...state,
+				[ action.importerId ]: {
+					...state[ action.importerId ],
+					importerState: appStates.UPLOAD_FAILURE,
+					errorData: { type: 'uploadError', description: action.error },
+				},
+			};
+
+		case IMPORTS_UPLOAD_COMPLETED:
+			return {
+				...omit( state, action.importerId ),
+				[ action.importerStatus.importerId ]: action.importerStatus,
+			};
+
+		case IMPORTS_IMPORT_RECEIVE:
+			// The REST endpoint returns an empty `[]` array if there are no known imports.
+			// In that case we don't update the `importerStatus` map, and only mark it as hydrated.
+			if ( isEmpty( action.importerStatus ) ) {
+				return state;
+			}
+
+			return omitBy(
+				{
+					// filter the original set of importers...
+					...state,
+					// ...and the importer being received.
+					[ action.importerStatus.importerId ]: fromApi( action.importerStatus ),
+				},
+				( importer ) =>
+					[ appStates.CANCEL_PENDING, appStates.DEFUNCT, appStates.EXPIRED ].includes(
```

Comment:
This could actually be a small standalone PR. The story is that on my account, the `/imports` endpoint returns a very old WIX import (I did it more than a year ago) that has an `importExpired` state. We want to ignore these imports and act like they didn't exist.

Now because `importExpired` is not defined in `importerStateMap`, trying to map it crashes the `apiToAppState`  function. The exception used to be swallowed by the `.catch( apiFailure )` handler, and the expired importer was rightfully ignored 🙂 After the Redux refactor, this hidden bug became apparent.